### PR TITLE
Compare constraints lazily across partial type declarations

### LINKED ISSUE #34841: StackOverflow with latest compiler when enabling nullable on a specific generic file
**Version Used**: 3.100.19.20601 (f50c04c1)

**Steps to Reproduce**:

1. Adding `#nullable enable` in this file: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.CoreCLR.cs
2. build coreclr with `build.cmd -windowscorlib`

**Expected Behavior**:
Compiles fine
**Actual Behavior**:
Crashes with: StackOverflowException

**Call stack**: 
[nullable-generic-stack.txt](https://github.com/dotnet/roslyn/files/3056163/nullable-generic-stack.txt)
 
I do have a crash dump but it is too big to upload, I can share it offline if needed.

cc: @jaredpar @cston @dotnet/nullablefc

### PR DESCRIPTION
**Pull request by cston**
Fixes #34841.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 1]
```diff
@@ -319,6 +309,36 @@ private ImmutableArray<TypeParameterConstraintClause> MakeTypeParameterConstrain
             return results;
         }
 
+        private ImmutableArray<TypeParameterConstraintClause> MakeTypeParameterConstraintsLate(
+            ImmutableArray<TypeParameterConstraintClause> constraintClauses,
+            DiagnosticBag diagnostics)
+        {
+            var typeParameters = TypeParameters;
+
+            if (constraintClauses.Length > 0)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: >if (constraintClauses.Length > 0) [](start = 12, length = 33)

Is this check redundant?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 2]
```diff
@@ -319,6 +309,36 @@ private ImmutableArray<TypeParameterConstraintClause> MakeTypeParameterConstrain
             return results;
         }
 
+        private ImmutableArray<TypeParameterConstraintClause> MakeTypeParameterConstraintsLate(
+            ImmutableArray<TypeParameterConstraintClause> constraintClauses,
+            DiagnosticBag diagnostics)
+        {
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: >{ [](start = 8, length = 1)

Consider adding an assert ```constraintClauses.IsEarly()```

**Comment by AlekseyTs** [comment]
Could you provide some details on what was the root cause of the stack overflow?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

**Comment by cston** [comment]
`HaveSameConstraints()` compares `IEquatable<T>` with `IEquatable<T>` which calls `TypeWithAnnotations.Equals()` with `T` and `T` which calls `IsValueType` on each `TypeParameter` which gets the constraints.

---
In reply to: [481044805](https://github.com/dotnet/roslyn/pull/34850#issuecomment-481044805) [](ancestors = 481044805)

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -87428,5 +87428,114 @@ static void M(string? x)
...
+{
+}";
+            var source2 =
+@"using System;
+partial class C<T> where T : IEquatable<T>
+{
+}";
+            var comp = CreateCompilation(new[] { source1, source2 });
+            comp.VerifyDiagnostics(
+                // (3,15): error CS0265: Partial declarations of 'C<T>' have inconsistent constraints for type parameter 'T'
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Probably unrelated, but this is a terrible error message. We should say _what_ the inconsistencies are.

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: It looks like [https://github.com/dotnet/roslyn/issues/30229](https://github.com/dotnet/roslyn/issues/30229) is tracking this diagnostic.

---
In reply to: [273679942](https://github.com/dotnet/roslyn/pull/34850#discussion_r273679942) [](ancestors = 273679942)

**Pull request merged**